File: src/main/java/engineer/thomas_werner/euler/math/Factorizer.java
package engineer.thomas_werner.euler.math;

import java.util.Collection;
import java.util.LinkedList;
import java.util.List;

public class Factorizer {

    public void getPrimeFactors(long number, Collection<Long> factors) {
        if (0 == number % 2) {
            factors.add(2L);
            getPrimeFactors(number / 2L, factors);
        }

        boolean found = false;
        for(long l=3; l*l <= number; l+=2) {
            if(0 == number % l) {
                getPrimeFactors(l, factors);
                getPrimeFactors(number / l, factors);
                found = true;
                break;
            }
        }

        if(!found)
            factors.add(number);
    }

    public List<Long> getDivisors(long number) {
        final List<Long> factors = new LinkedList<>();
        for(long l=1; l < number; l++)
            if(0 == number % l)
                factors.add(l);
        return factors;
    }

}

File: src/main/kotlin/de/huddeldaddel/euler/math/Factorizer.kt
package de.huddeldaddel.euler.math

import java.util.LinkedList

class Factorizer {

    fun getPrimeFactors(number: Long, factors: MutableCollection<Long>) {
        if (0L == number % 2) {
            factors.add(2L)
            getPrimeFactors(number / 2L, factors)
        }

        var found = false
        var l: Long = 3
        while (l * l <= number) {
            if (0L == number % l) {
                getPrimeFactors(l, factors)
                getPrimeFactors(number / l, factors)
                found = true
                break
            }
            l += 2
        }

        if (!found)
            factors.add(number)
    }

    fun getDivisors(number: Long): List<Long> {
        val factors = LinkedList<Long>()
        for (l in 1 until number)
            if (0L == number % l)
                factors.add(l)
        return factors
    }

}
